File: week 6 - Python/credit.py
from cs50 import get_string

card_number = get_string("Number :") # get user input

total_sum = 0 # assign total digit sum

# iterate over every digit with its index
for ind, number in enumerate(reversed(card_number)):
    # Add the digits from first digit by twos to total_sum
    if ind % 2 == 0:
        total_sum += int(number)
    # Multiply every other digit by 2, starting with the number’s second-to-last digit, add to total_sum
    else:
        number = int(number) * 2
        # If multiplied number is bigger than 9, get sum of its digits
        if number > 9:
            total_sum += int((number % 100) / 10) + (number % 10)
        else:
            total_sum += number

# if total_sum is the power of 10, it is Valid card number
if total_sum % 10 != 0:
    print("INVALID\n")

# Check Visa, Amex or Mastercard
else:
    if (len(card_number) == 13 or len(card_number) == 16) and card_number[0] == "4":
        print("VISA\n")

    if len(card_number) == 16 and (card_number[0:2] == "51" or card_number[0:2] == "52" or card_number[0:2] == "53" or card_number[0:2] == "54"  or card_number[0:2] == "55"):
        print("MASTERCARD\n")

    if len(card_number) == 15 and (card_number[0:2] == '34' or card_number[0:2] == '37'):
        print("AMEX\n")

File: CS50: Introduction to Computer Science/week 1 - C/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long card_number;
    int c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, sum, length = 0, last2_digits;

    do
    {
        card_number = get_long("Number: ");
    }
    while (card_number < 0); //Take card number

    //Get the digits which are multiplied by 2
    c1 = ((card_number % 100) / 10) * 2;
    c3 = ((card_number % 10000) / 1000) * 2;
    c5 = ((card_number % 1000000) / 100000) * 2;
    c7 = ((card_number % 100000000) / 10000000) * 2;
    c9 = ((card_number % 10000000000) / 1000000000) * 2;
    c11 = ((card_number % 1000000000000) / 100000000000) * 2;
    c13 = ((card_number % 100000000000000) / 10000000000000) * 2;
    c15 = ((card_number % 10000000000000000) / 1000000000000000) * 2;

    //Calculate the digit sum for each digit
    c1 = (c1 % 100) / 10 + c1 % 10;
    c3 = (c3 % 100) / 10 + c3 % 10;
    c5 = (c5 % 100) / 10 + c5 % 10;
    c7 = (c7 % 100) / 10 + c7 % 10;
    c9 = (c9 % 100) / 10 + c9 % 10;
    c11 = (c11 % 100) / 10 + c11 % 10;
    c13 = (c13 % 100) / 10 + c13 % 10;
    c15 = (c15 % 100) / 10 + c15 % 10;

    //Get the digits which are not multiplied by 2
    c0 = card_number % 10;
    c2 = (card_number % 1000) / 100;
    c4 = (card_number % 100000) / 10000;
    c6 = (card_number % 10000000) / 1000000;
    c8 = (card_number % 1000000000) / 100000000;
    c10 = (card_number % 100000000000) / 10000000000;
    c12 = (card_number % 10000000000000) / 1000000000000;
    c14 = (card_number % 1000000000000000) / 100000000000000;

    //Calculate the total sum
    sum = c0 + c1 + c2 + c3 + c4 + c5 + c6 + c7 + c8 + c9 + c10 + c11 + c12 + c13 + c14 + c15;

    printf("%d\n", sum);

    if (sum % 10 != 0) //Check Valid or Invalid?
    {
        printf("INVALID\n");
    }

    else //Check Visa or Amex or MasterCard?
    {
        while (card_number > 0) //Find length and last 2 digits
        {
            card_number = card_number / 10;

            if (card_number<100 & card_number>10)
            {
                last2_digits = card_number;
            }

            length++;
        }


        if ((length == 13 || length == 16) && last2_digits / 10 == 4)
        {
            printf("VISA\n");
        }

        if (length == 16 && (last2_digits == 51 || last2_digits == 52 || last2_digits == 53 || last2_digits == 54  || last2_digits == 55))
        {
            printf("MASTERCARD\n");
        }

        if (length == 15 && (last2_digits == 34 || last2_digits == 37))
        {
            printf("AMEX\n");
        }


    }
}